packages/block-editor/src/components/list-view/index.js

```diff
@@ -89,11 +98,64 @@ function ListView(
 		[ draggedClientIds ]
 	);
 	const selectEditorBlock = useCallback(
-		( clientId ) => {
-			selectBlock( clientId );
+		async ( clientId, event ) => {
+			if ( ! event.shiftKey ) {
+				await clearSelectedBlock();
+				selectBlock( clientId, -1 );
+			} else if ( event.shiftKey ) {
+				// To handle multiple block selection via the `SHIFT` key, prevent
+				// the browser default behavior of opening the link in a new window.
+				event.preventDefault();
+
+				// Select a single block if no block is selected yet.
+				if ( ! hasSelectedBlock() && ! hasMultiSelection() ) {
+					selectBlock( clientId, -1 );
+					return;
+				}
+
+				const blockSelectionStart = getBlockSelectionStart();
+
+				// By checking `blockSelectionStart` to be set, we handle the
+				// case where we select a single block. We also have to check
+				// the selectionEnd (clientId) not to be included in the
+				// `blockSelectionStart`'s parents because the click event is
+				// propagated.
+				const startParents = getBlockParents( blockSelectionStart );
+
+				if (
+					blockSelectionStart &&
+					blockSelectionStart !== clientId &&
+					! startParents?.includes( clientId )
+				) {
+					const startPath = [ ...startParents, blockSelectionStart ];
+					const endPath = [
+						...getBlockParents( clientId ),
+						clientId,
+					];
+					const depth =
+						Math.min( startPath.length, endPath.length ) - 1;
+					const start = startPath[ depth ];
+					const end = endPath[ depth ];
```

Comment:
🤔 It may be worth decomposing this a bit and unit testing the calculation of start and end. We probably need to document a few corner cases with selections across nested children.